From 4bc733552be725fa5245d78568f370c27136769c Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:08:33 -0400
Subject: [usb] option: fix runtime PM handling

Message-id: <1410887463-20674-442-git-send-email-dzickus@redhat.com>
Patchwork-id: 95601
O-Subject: [RHEL7 PATCH 441/591] USB: option: fix runtime PM handling
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit acf47d4f9c39b1cba467aa9442fc2efe0b1da741
Author: Johan Hovold <jhovold@gmail.com>
Date:   Mon May 26 19:23:10 2014 +0200

    USB: option: fix runtime PM handling

    Fix potential I/O while runtime suspended due to missing PM operations
    in send_setup.

    Fixes: 383cedc3bb43 ("USB: serial: full autosuspend support for the
    option driver")

    Cc: <stable@vger.kernel.org>        # v2.6.32
    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/option.c b/drivers/usb/serial/option.c
index f213ee9..802c069 100644
--- a/drivers/usb/serial/option.c
+++ b/drivers/usb/serial/option.c
@@ -1923,6 +1923,7 @@ static int option_send_setup(struct usb_serial_port *port)
  struct option_private *priv = intfdata->private;
  struct usb_wwan_port_private *portdata;
  int val = 0;
+ int res;
 
  portdata = usb_get_serial_port_data(port);
 
@@ -1931,9 +1932,17 @@ static int option_send_setup(struct usb_serial_port *port)
  if (portdata->rts_state)
   val |= 0x02;
 
- return usb_control_msg(serial->dev, usb_rcvctrlpipe(serial->dev, 0),
+ res = usb_autopm_get_interface(serial->interface);
+ if (res)
+  return res;
+
+ res = usb_control_msg(serial->dev, usb_rcvctrlpipe(serial->dev, 0),
     0x22, 0x21, val, priv->bInterfaceNumber, NULL,
     0, USB_CTRL_SET_TIMEOUT);
+
+ usb_autopm_put_interface(serial->interface);
+
+ return res;
 }
 
 MODULE_AUTHOR(DRIVER_AUTHOR);
-- 
1.7.1